File: dx/src/uipp/windows/XmlPreferences.cpp
/***********************************************************************/
/* Open Visualization Data Explorer                                    */
/* (C) Copyright IBM Corp. 2004                                        */
/* ALL RIGHTS RESERVED                                                 */
/* This code licensed under the                                        */
/*    "IBM PUBLIC LICENSE - Open Visualization Data Explorer"          */
/***********************************************************************/

#include <dxconfig.h>
#include "defines.h"

#include "XmlPreferences.h"

#using <System.Xml.dll>
using namespace System;
using namespace System::Xml;
using namespace System::Runtime::InteropServices;


XmlPreferences* theXmlPreferences = 0;

void XmlPreferences::BuildTheXmlPreferences()
{
    if (theXmlPreferences == 0) 
	theXmlPreferences = new XmlPreferences();
}

XmlPreferences::XmlPreferences() {
	this->prefs = new XmlDocument();
	this->root = NULL;
	this->fileExisted = false;
}

XmlPreferences::~XmlPreferences() {

}

bool XmlPreferences::readPrefs(const char *filename) {
	System::String __gc *s = new System::String(filename);
	XmlTextReader* reader = new XmlTextReader(s);

	if(! reader) {
		return false;
	}

	try{
		this->prefs->Load(reader);
		this->fileExisted = true;
	} catch(System::Xml::XmlException *e) {
		ASSERT("Corrupted prefs file--try deleting");
	}

	reader->Close();

	this->root = this->prefs->SelectSingleNode(S"PREFS");
	if(!this->root) {
		XmlElement *elem = this->prefs->CreateElement(S"PREFS");
		this->prefs->AppendChild(elem);
		this->root = elem;
	}

	return this->fileExisted;
}

bool XmlPreferences::createPrefs() {
	ASSERT(this->prefs);
	try {
		XmlElement *elem = this->prefs->CreateElement(S"PREFS");
		this->prefs->AppendChild(elem);
		this->root = elem;
	} catch(Exception *e) {
		return false;
	}
	return true;
}



bool XmlPreferences::writePrefs(const char *filename) {
	this->fileExisted = false;

	// TODO: May want to remove non-persistent prefs. If so,
	// will need to traverse tree and remove them before 
	// trying to write file.

	try{
		this->prefs->Save(filename);
		this->fileExisted = true;
	} catch(System::Xml::XmlException *e) {
		// couldn't write file
	}
	return this->fileExisted;
}

bool XmlPreferences::getPref(const char *prefName, char *&value) {
	if(!this->prefs)
		return false;

	// Xml is in the form.
	//<PREF Name='UIRoot'><TYPE>string</TYPE><VALUE>/usr/local/dx</VALUE></PREF>
	//<PREF Name='wizard'><TYPE>bool</TYPE><VALUE>false</VALUE></PREF>

	System::String __gc *lookup = S"DXUIPREF[@Name='";
	lookup = System::String::Concat(lookup, new System::String(prefName));
	lookup = System::String::Concat(lookup, S"']");
	System::String __gc *val;
	try {
		XmlNode *n = this->root->SelectSingleNode(lookup);

		if(!n)
			return false;
		XmlNode *n2 = n->SelectSingleNode(S"VALUE");
		val = n2->InnerText;
	} catch (System::Xml::XmlException *e) {
		ASSERT("Corrupted prefs file--try deleting");
	}

	if(val->Length == 0) {
		value = NULL;
		return true;
	}

	value = new char[val->Length + 1];
	strncpy(value, 
		(const char*) (Marshal::StringToHGlobalAnsi(val)).ToPointer(),
		val->Length);
	value[val->Length] = '\0';

	return true;
}

bool XmlPreferences::getPref(const char *prefName, bool &value) {
	if(!this->prefs)
		return false;

	// Xml is in the form.
	//<PREF Name='UIRoot'><TYPE>string</TYPE><VALUE>/usr/local/dx</VALUE></PREF>
	//<PREF Name='wizard'><TYPE>bool</TYPE><VALUE>false</VALUE></PREF>

	System::String __gc *lookup = S"DXUIPREF[@Name='";
	lookup = System::String::Concat(lookup, new System::String(prefName));
	lookup = System::String::Concat(lookup, S"']");
	System::String __gc *val;
	try {
		XmlNode *n = this->root->SelectSingleNode(lookup);

		if(!n)
			return false;
		XmlNode *n2 = n->SelectSingleNode(S"VALUE");
		val = n2->InnerText;
	} catch (System::Xml::XmlException *e) {
		ASSERT("Corrupted prefs file--try deleting");
	}

	if(val->CompareTo(S"true") == 0)
		value = true;
	else
		value = false;

	return true;
}

bool XmlPreferences::getPref(const char *prefName, int &value) {
	if(!this->prefs)
		return false;

	// Xml is in the form.
	//<PREF Name='UIRoot'><TYPE>string</TYPE><VALUE>/usr/local/dx</VALUE></PREF>
	//<PREF Name='wizard'><TYPE>bool</TYPE><VALUE>false</VALUE></PREF>

	System::String __gc *lookup = S"DXUIPREF[@Name='";
	lookup = System::String::Concat(lookup, new System::String(prefName));
	lookup = System::String::Concat(lookup, S"']");
	System::String __gc *val;
	try {
		XmlNode *n = this->root->SelectSingleNode(lookup);

		if(!n)
			return false;
		XmlNode *n2 = n->SelectSingleNode(S"VALUE");
		val = n2->InnerText;
	} catch (System::Xml::XmlException *e) {
		ASSERT("Corrupted prefs file--try deleting");
	}

	value = Convert::ToInt32(val);

	return true;
}


bool XmlPreferences::getPref(const char* prefName, PrefType& type, char *&value) {
	if(!this->prefs)
		return false;

	// Xml is in the form.
	//<PREF Name='UIRoot'><TYPE>string</TYPE><VALUE>/usr/local/dx</VALUE></PREF>
	//<PREF Name='wizard'><TYPE>bool</TYPE><VALUE>false</VALUE></PREF>
	System::String __gc *lookup = S"DXUIPREF[@Name='";
	lookup = System::String::Concat(lookup, new System::String(prefName));
	lookup = System::String::Concat(lookup, S"']");

	System::String __gc *val, *typeS;
	try {
		XmlNode *n = this->root->SelectSingleNode(lookup);
		if(!n)
			return false;

		XmlNode *n2 = n->SelectSingleNode(S"VALUE");
		val = n2->InnerText;

		n2 = n->SelectSingleNode(S"TYPE");
		typeS = n2->InnerText;

	} catch (System::Xml::XmlException *e) {
		printf("Corrupted prefs file--try deleting\n");
		return false;
	}

	if(val->Length == 0) {
		value = NULL;
		return true;
	}

	value = new char[val->Length + 1];
	strncpy(value, 
		(const char*) (Marshal::StringToHGlobalAnsi(val)).ToPointer(),
		val->Length);
	value[val->Length] = '\0';

	if(typeS->CompareTo(S"bool") == 0)
		type = TypeBool;
	else if(typeS->CompareTo(S"int") == 0)
		type = TypeInt;
	else if(typeS->CompareTo(S"float") == 0)
		type = TypeFloat;
	else
		type = TypeString;

	return true;

}

bool XmlPreferences::setPref(const char *prefName, const char *value, bool persist) {
	if(!this->prefs)
		return false;
	System::String __gc *lookup = S"DXUIPREF[@Name='";
	lookup = System::String::Concat(lookup, new System::String(prefName));
	lookup = System::String::Concat(lookup, S"']");

	// First see if it exists. If so use replace.
	try {
		XmlNode *n = this->root->SelectSingleNode(lookup);
		if(n) {
			try{
				XmlElement *elem = this->prefs->CreateElement(S"DXUIPREF");
				elem->SetAttribute(S"Name", new System::String(prefName));
				XmlElement *val = this->prefs->CreateElement(S"VALUE");
				val->InnerText = new System::String(value);
				elem->AppendChild(val);
				XmlElement *persistent = this->prefs->CreateElement(S"PERSISTENT");
				if(persist)
					persistent->InnerText = S"true";
				else
					persistent->InnerText = S"false";
				elem->AppendChild(persistent);
				this->root->ReplaceChild(elem, n);
			} catch (System::Xml::XmlException *e) {
				return false;
			}
		} else
			try{
				XmlElement *elem = this->prefs->CreateElement(S"DXUIPREF");
				elem->SetAttribute(S"Name", new System::String(prefName));
				XmlElement *val = this->prefs->CreateElement(S"VALUE");
				val->InnerText = new System::String(value);
				elem->AppendChild(val);
				XmlElement *persistent = this->prefs->CreateElement(S"PERSISTENT");
				if(persist)
					persistent->InnerText = S"true";
				else
					persistent->InnerText = S"false";
				elem->AppendChild(persistent);
				this->root->AppendChild(elem);
			} catch (System::Xml::XmlException *e) {
				return false;
			}
	} catch (System::Xml::XmlException *e) {
		return false;
	}

	return true;

}

bool XmlPreferences::setPref(const char *prefName, PrefType type, const char *value, bool persist) {
	if(!this->prefs)
		return false;
	System::String __gc *lookup = S"DXUIPREF[@Name='";
	lookup = System::String::Concat(lookup, new System::String(prefName));
	lookup = System::String::Concat(lookup, S"']");

	System::String __gc *typeS;
	switch(type) {
		case TypeString:
			typeS = new System::String(S"string");
			break;
		case TypeInt:
			typeS = new System::String(S"int");
			break;
		case TypeBool:
			typeS = new System::String(S"bool");
			break;
		case TypeFloat:
			typeS = new System::String(S"float");
			break;
	}

	// First see if it exists. If so use replace.
	try {
		XmlNode *n = this->root->SelectSingleNode(lookup);
		if(n) {
			try{
				XmlElement *elem = this->prefs->CreateElement(S"DXUIPREF");
				elem->SetAttribute(S"Name", new System::String(prefName));
				XmlElement *val = this->prefs->CreateElement(S"VALUE");
				val->InnerText = new System::String(value);
				elem->AppendChild(val);
				XmlElement *tp = this->prefs->CreateElement(S"TYPE");
				tp->InnerText = typeS;
				elem->AppendChild(tp);
				XmlElement *persistent = this->prefs->CreateElement(S"PERSISTENT");
				if(persist)
					persistent->InnerText = S"true";
				else
					persistent->InnerText = S"false";
				elem->AppendChild(persistent);
				this->root->ReplaceChild(elem, n);
			} catch (System::Xml::XmlException *e) {
				return false;
			}
		} else
			try{
				XmlElement *elem = this->prefs->CreateElement(S"DXUIPREF");
				elem->SetAttribute(S"Name", new System::String(prefName));
				XmlElement *val = this->prefs->CreateElement(S"VALUE");
				val->InnerText = new System::String(value);
				elem->AppendChild(val);
				XmlElement *tp = this->prefs->CreateElement(S"TYPE");
				tp->InnerText = typeS;
				elem->AppendChild(tp);
				XmlElement *persistent = this->prefs->CreateElement(S"PERSISTENT");
				if(persist)
					persistent->InnerText = S"true";
				else
					persistent->InnerText = S"false";
				elem->AppendChild(persistent);
				this->root->AppendChild(elem);
			} catch (System::Xml::XmlException *e) {
				return false;
			}
	} catch (System::Xml::XmlException *e) {
		return false;
	}

	return true;

}

bool XmlPreferences::setDefault(const char *prefName, const char *value) {
	if(!this->prefs)
		return false;
	System::String __gc *lookup = S"DXUIPREF[@Name='";
	lookup = System::String::Concat(lookup, new System::String(prefName));
	lookup = System::String::Concat(lookup, S"']");

	// First see if it exists. If so just return true.
	try {
		XmlNode *n = this->root->SelectSingleNode(lookup);
		if(n) {
			return true;
		} else
			try{
				XmlElement *elem = this->prefs->CreateElement(S"DXUIPREF");
				elem->SetAttribute(S"Name", new System::String(prefName));
				XmlElement *val = this->prefs->CreateElement(S"VALUE");
				val->InnerText = new System::String(value);
				elem->AppendChild(val);
				XmlElement *persistent = this->prefs->CreateElement(S"PERSISTENT");
				persistent->InnerText = false;
				elem->AppendChild(persistent);
				this->root->AppendChild(elem);
			} catch (System::Xml::XmlException *e) {
				return false;
			}
	} catch (System::Xml::XmlException *e) {
		return false;
	}
	return true;
}

bool XmlPreferences::setDefault(const char *prefName, PrefType type, const char *value) {
	if(!this->prefs)
		return false;
	System::String __gc *lookup = S"DXUIPREF[@Name='";
	lookup = System::String::Concat(lookup, new System::String(prefName));
	lookup = System::String::Concat(lookup, S"']");

	System::String __gc *typeS;
	switch(type) {
		case TypeString:
			typeS = new System::String(S"string");
			break;
		case TypeInt:
			typeS = new System::String(S"int");
			break;
		case TypeBool:
			typeS = new System::String(S"bool");
			break;
		case TypeFloat:
			typeS = new System::String(S"float");
			break;
	}

	// First see if it exists. If so just return.
	try {
		XmlNode *n = this->root->SelectSingleNode(lookup);
		if(n) {
			return true;	
		} else
			try{
				XmlElement *elem = this->prefs->CreateElement(S"DXUIPREF");
				elem->SetAttribute(S"Name", new System::String(prefName));
				XmlElement *val = this->prefs->CreateElement(S"VALUE");
				val->InnerText = new System::String(value);
				elem->AppendChild(val);
				XmlElement *tp = this->prefs->CreateElement(S"TYPE");
				tp->InnerText = typeS;
				elem->AppendChild(tp);
				XmlElement *persistent = this->prefs->CreateElement(S"PERSISTENT");
				persistent->InnerText = S"false";
				elem->AppendChild(persistent);
				this->root->AppendChild(elem);
			} catch (System::Xml::XmlException *e) {
				return false;
			}
	} catch (System::Xml::XmlException *e) {
		return false;
	}

	return true;

}

File: VisualDX/WinDXui/XmlPreferences.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using System.Diagnostics;

namespace WinDX.UI
{
    public class XmlPreferences
    {
        public enum PrefType { TypeString, TypeInt, TypeBool, TypeFloat };

        private XmlDocument prefs;
        private bool fileExisted = false;
        private XmlNode root = null;
        private const int MaxStoredList = 10;

        public XmlPreferences()
        {
            prefs = new XmlDocument();
        }

        public bool ReadPrefs(String filename)
        {
            XmlTextReader reader = new XmlTextReader(filename);

            if (reader == null)
            {
                return false;
            }
            try
            {
                prefs.Load(reader);
                fileExisted = true;
            }
            catch (XmlException)
            {
                Console.WriteLine("Corrupted prefs file--try deleting");
            }

            reader.Close();
            root = prefs.SelectSingleNode("PREFS");
            if (root == null)
            {
                XmlElement elem = prefs.CreateElement("PREFS");
                prefs.AppendChild(elem);
                root = elem;
            }
            return fileExisted;
        }

        public bool CreatePrefs()
        {
            Debug.Assert(prefs != null);
            try
            {
                XmlElement elem = prefs.CreateElement("PREFS");
                prefs.AppendChild(elem);
                root = elem;
            }
            catch (Exception)
            {
                return false;
            }
            return true;
        }

        public bool WritePrefs(String filename)
        {
            fileExisted = false;

            // Find only Persistent nodes and write them out to the prefs
            // file. This is done by creating a new XmlDocument, importing
            // appropriate nodes and adding them before writing.

            try
            {
                String lookup = "DXUIPREF[@Persistent='true']";
                XmlNodeList nl = root.SelectNodes(lookup);

                if (nl.Count > 0)
                {
                    XmlDocument doc = new XmlDocument();
                    XmlElement elem = doc.CreateElement("PREFS");
                    doc.AppendChild(elem);
                    for (int i = 0; i < nl.Count; i++)
                        elem.AppendChild(doc.ImportNode(nl.Item(i), true));

                    doc.Save(filename);
                    fileExisted = true;
                }
            }
            catch (XmlException)
            {
                Console.WriteLine("Couldn't write file--check permissions.");
            }
            return fileExisted;
        }

        public bool GetPref(String prefName, out String value)
        {
            if (root == null)
            {
                value = null;
                return false;
            }

            value = "";
            if (prefs == null)
                return false;

            // Xml is in the form.
            //<PREF Name='UIRoot' Persistent="true"><TYPE>string</TYPE><VALUE>/usr/local/dx</VALUE></PREF>
            //<PREF Name='wizard' Persistent="false"><TYPE>bool</TYPE><VALUE>false</VALUE></PREF>

            String lookup = "DXUIPREF[@Name='";
            lookup += prefName;
            lookup += "']";

            String val = null;

            try
            {
                XmlNode n = root.SelectSingleNode(lookup);

                if (n == null)
                    return false;

                XmlNode n2 = n.SelectSingleNode("VALUE");
                val = n2.InnerText;
            }
            catch (XmlException)
            {
                Console.WriteLine("Corrupted prefs file--try deleting it");
            }

            if (val.Length == 0)
            {
                value = null;
                return true;
            }
            value = val;
            return true;
        }

        public bool GetPref(String prefName, out bool value)
        {
            if (root == null)
            {
                value = false;
                return false;
            }

            value = false;
            if (prefs == null)
                return false;

            // Xml is in the form.
            //<PREF Name='UIRoot' Persistent="true"><TYPE>string</TYPE><VALUE>/usr/local/dx</VALUE></PREF>
            //<PREF Name='wizard' Persistent="false"><TYPE>bool</TYPE><VALUE>false</VALUE></PREF>

            String lookup = "DXUIPREF[@Name='";
            lookup += prefName;
            lookup += "']";

            String val = null;

            try
            {
                XmlNode n = root.SelectSingleNode(lookup);

                if (n == null)
                    return false;

                XmlNode n2 = n.SelectSingleNode("VALUE");
                val = n2.InnerText;
            }
            catch (XmlException)
            {
                Console.WriteLine("Corrupted prefs file--try deleting it");
            }

            if (val.CompareTo("true") == 0)
                value = true;
            else
                value = false;

            return true;
        }

        public bool GetPref(String prefName, out int value)
        {
            value = 0;

            if (prefs == null)
                return false;

            // Xml is in the form.
            //<PREF Name='UIRoot' Persistent="true"><TYPE>string</TYPE><VALUE>/usr/local/dx</VALUE></PREF>
            //<PREF Name='wizard' Persistent="false"><TYPE>bool</TYPE><VALUE>false</VALUE></PREF>

            String lookup = "DXUIPREF[@Name='";
            lookup += prefName;
            lookup += "']";

            String val = null;

            try
            {
                XmlNode n = root.SelectSingleNode(lookup);

                if (n == null)
                    return false;

                XmlNode n2 = n.SelectSingleNode("VALUE");
                val = n2.InnerText;
            }
            catch (XmlException)
            {
                Console.WriteLine("Corrupted prefs file--try deleting it");
            }

            value = Int32.Parse(val);

            return true;
        }

        public bool GetPref(String prefName, out PrefType type, out String value)
        {
            type = PrefType.TypeString;
            value = "";
            if (prefs == null)
                return false;

            // Xml is in the form.
            //<PREF Name='UIRoot' Persistent="true"><TYPE>string</TYPE><VALUE>/usr/local/dx</VALUE></PREF>
            //<PREF Name='wizard' Persistent="false"><TYPE>bool</TYPE><VALUE>false</VALUE></PREF>

            String lookup = "DXUIPREF[@Name='";
            lookup += prefName;
            lookup += "']";

            String val = null, typeS = null;

            try
            {
                XmlNode n = root.SelectSingleNode(lookup);

                if (n == null)
                    return false;

                XmlNode n2 = n.SelectSingleNode("VALUE");
                val = n2.InnerText;

                n2 = n.SelectSingleNode("TYPE");
                typeS = n2.InnerText;
            }
            catch (XmlException)
            {
                Console.WriteLine("Corrupted prefs file--try deleting it");
            }

            if (val.Length == 0)
            {
                value = null;
                return true;
            }

            value = val;
            if (typeS.CompareTo("bool") == 0)
                type = PrefType.TypeBool;
            else if (typeS.CompareTo("int") == 0)
                type = PrefType.TypeInt;
            else if (typeS.CompareTo("float") == 0)
                type = PrefType.TypeFloat;
            else
                type = PrefType.TypeString;

            return true;
        }

        /// <summary>
        /// Get a value for a specific preference name.
        /// </summary>
        /// <param name="prefName">the param name to query</param>
        /// <param name="value">sets value to a list of strings for the pref</param>
        /// <returns>true if pref found, false if not found</returns>
        public bool GetPref(String prefName, out List<String> value)
        {
            String val;
            bool gotPref = GetPref(prefName, out val);
            if (!gotPref)
            {
                value = null;
                return false;
            }
            //Now break the string into a list
            value = Utils.StringTokenizer(val, "|", null);
            return true;
        }

        /// <summary>
        /// Add a preference value to the list of values all ready existing.
        /// </summary>
        /// <param name="prefName"></param>
        /// <param name="value"></param>
        /// <param name="persist"></param>
        /// <returns></returns>
        public bool AddListPref(String prefName, String value) { return AddListPref(prefName, value, true); }
        public bool AddListPref(String prefName, String value, bool persist)
        {
            String curPrefVal;
            if (GetPref(prefName, out curPrefVal))
            {
                List<String> valueL = Utils.StringTokenizer(curPrefVal, "|", null);
                valueL.Add(value);
                String newval = "";
                for (int i = 0; i < valueL.Count - 1; i++)
                    newval += valueL[i] + "|";
                newval += valueL[valueL.Count - 1];
                return SetPref(prefName, newval, persist);
            }
            else
                return SetPref(prefName, value, persist);
        }

        public bool SetPref(String prefName, String value, bool persist)
        {
            if (prefs == null)
            {
                return false;
            }

            String lookup = "DXUIPREF[@Name='" + prefName + "']";

            try
            {
                XmlNode n = root.SelectSingleNode(lookup);
                if (n != null)
                {
                    try
                    {
                        XmlElement elem = prefs.CreateElement("DXUIPREF");
                        elem.SetAttribute("Name", prefName);
                        if (persist)
                            elem.SetAttribute("Persistent", "true");
                        else
                            elem.SetAttribute("Persistent", "false");
                        XmlElement val = prefs.CreateElement("VALUE");
                        val.InnerText = value;
                        elem.AppendChild(val);
                        root.ReplaceChild(elem, n);
                    }
                    catch (XmlException)
                    {
                        return false;
                    }
                }
                else
                    try
                    {
                        XmlElement elem = prefs.CreateElement("DXUIPREF");
                        elem.SetAttribute("Name", prefName);
                        if (persist)
                            elem.SetAttribute("Persistent", "true");
                        else
                            elem.SetAttribute("Persistent", "false");
                        XmlElement val = prefs.CreateElement("VALUE");
                        val.InnerText = value;
                        elem.AppendChild(val);
                        root.AppendChild(elem);
                    }
                    catch (XmlException)
                    {
                        return false;
                    }
            }
            catch (XmlException)
            {
                return false;
            }

            return true;
        }

        public bool SetPref(String prefName, PrefType type, String value, bool persist)
        {
            if (prefs == null)
                return false;

            String lookup = "DXUIPREF[@Name='" + prefName + "']";

            String typeS = null;
            switch (type)
            {
                case PrefType.TypeString:
                    typeS = "string";
                    break;
                case PrefType.TypeInt:
                    typeS = "int";
                    break;
                case PrefType.TypeBool:
                    typeS = "bool";
                    break;
                case PrefType.TypeFloat:
                    typeS = "float";
                    break;
            }

            try
            {
                XmlNode n = root.SelectSingleNode(lookup);
                if (n != null)
                {
                    try
                    {
                        XmlElement elem = prefs.CreateElement("DXUIPREF");
                        elem.SetAttribute("Name", prefName);
                        if (persist)
                            elem.SetAttribute("Persistent", "true");
                        else
                            elem.SetAttribute("Persistent", "false");
                        XmlElement val = prefs.CreateElement("VALUE");
                        val.InnerText = value;
                        elem.AppendChild(val);
                        XmlElement tp = prefs.CreateElement("TYPE");
                        tp.InnerText = typeS;
                        elem.AppendChild(tp);
                        root.ReplaceChild(elem, n);
                    }
                    catch (XmlException)
                    {
                        return false;
                    }
                }
                else
                    try
                    {
                        XmlElement elem = prefs.CreateElement("DXUIPREF");
                        elem.SetAttribute("Name", prefName);
                        if (persist)
                            elem.SetAttribute("Persistent", "true");
                        else
                            elem.SetAttribute("Persistent", "false");
                        XmlElement val = prefs.CreateElement("VALUE");
                        val.InnerText = value;
                        elem.AppendChild(val);
                        XmlElement tp = prefs.CreateElement("TYPE");
                        tp.InnerText = typeS;
                        elem.AppendChild(tp);
                        root.AppendChild(elem);
                    }
                    catch (XmlException)
                    {
                        return false;
                    }
            }
            catch (XmlException)
            {
                return false;
            }
            return true;
        }

        public bool SetDefault(String prefName, String value)
        {
            if (prefs == null)
                return false;
            String lookup = "DXUIPREF[@Name='" + prefName + "']";

            try
            {
                XmlNode n = root.SelectSingleNode(lookup);
                if (n != null)
                    return true;
                try
                {
                    XmlElement elem = prefs.CreateElement("DXUIPREF");
                    elem.SetAttribute("Name", prefName);
                    elem.SetAttribute("Persistent", "false");
                    XmlElement val = prefs.CreateElement("VALUE");
                    val.InnerText = value;
                    elem.AppendChild(val);
                    root.AppendChild(elem);
                }
                catch (XmlException)
                {
                    return false;
                }
            }
            catch (XmlException)
            {
                return false;
            }
            return true;
        }

        /// <summary>
        /// Set a Default for a Pref. If it already exists, just return.
        /// In order to set a pref persistently (saves out to file) use
        /// SetPref.
        /// </summary>
        /// <param name="prefName"></param>
        /// <param name="type"></param>
        /// <param name="value"></param>
        /// <returns></returns>
        public bool SetDefault(String prefName, PrefType type, String value)
        {
            return SetDefault(prefName, type, value, false);
        }
        public bool SetDefault(String prefName, PrefType type, String value, bool persistent)
        {
            if (prefs == null)
                return false;
            String lookup = "DXUIPREF[@Name='" + prefName + "']";
            String typeS = null;
            switch (type)
            {
                case PrefType.TypeString:
                    typeS = "string";
                    break;
                case PrefType.TypeInt:
                    typeS = "int";
                    break;
                case PrefType.TypeBool:
                    typeS = "bool";
                    break;
                case PrefType.TypeFloat:
                    typeS = "float";
                    break;
            }

            // First see if it exists. If so just return.
            try
            {
                XmlNode n = root.SelectSingleNode(lookup);
                if (n != null)
                    return true;
                try
                {
                    XmlElement elem = prefs.CreateElement("DXUIPREF");
                    elem.SetAttribute("Name", prefName);
                    if(persistent)
                        elem.SetAttribute("Persistent", "true");
                    else
                        elem.SetAttribute("Persistent", "false");
                    XmlElement val = prefs.CreateElement("VALUE");
                    val.InnerText = value;
                    elem.AppendChild(val);
                    XmlElement tp = prefs.CreateElement("TYPE");
                    tp.InnerText = typeS;
                    elem.AppendChild(tp);
                    root.AppendChild(elem);
                }
                catch (XmlException)
                {
                    return false;
                }
            }
            catch (XmlException)
            {
                return false;
            }
            return true;
        }
        public static XmlPreferences theXmlPreferences = new XmlPreferences();
    }
}